### PR TITLE
Add update item endpoint

### PR DESCRIPTION
**Pull request by samstradling**
Adds the endpoint to update items.

### CODE REVIEW

File: handlers/patchItem_test.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+		expectedRes        interface{}
+		expectedStatusCode int
+		shouldCallDB       bool
+	}{
+		{
+			name:               "Returns 'Bad Request' when the path is empty",
+			path:               "",
+			listID:             "test-list-id",
+			output:             nil,
+			outputErr:          nil,
```
Comment (alexNaylorTeece) [inline handlers/patchItem_test.go:128]: If `shouldCallDB` is false, you don't need to specify these

If you wanted, you could have a struct defined above `tests` that was something like:
```
type mockUpdateItem struct {
    output data.Item
    err error
}
```
and then in `tests` you could have `mockUpdateItem *mockUpdateItem`
You could then use `tt.mockUpdateItem != nil` in place of `shouldCallDB`

Up to you though. Just different ways of doing same thing

**Review by alexNaylorTeece — CHANGES_REQUESTED** [review]

File: db/updateItem.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+	input := &dynamodb.PutItemInput{
+		Item:                itemToInsert,
+		TableName:           aws.String(d.conf.TableNames.Items),
+		ConditionExpression: aws.String("attribute_exists(Id) AND attribute_exists(ListId)"),
+	}
+
+	_, err = d.session.PutItem(input)
+
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (samstradling) [reply db/updateItem.go:30]: `err.type` doesn't seem to be a thing:
Screenshot 2020-11-05 at 08 34 02

File: handlers/patchItem.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		log.Printf("Error: %s", err.Error())
+		return nil, http.StatusInternalServerError
+	}
+
+	return item, http.StatusOK
+}
+
+func getName(body string) (*string, error) {
+	var item data.Item
+	err := json.Unmarshal([]byte(body), &item)
```
[the remark this answers is not in the document]
Comment (samstradling) [reply handlers/patchItem.go:60]: They could do it, but it'd have no effect, the same way that specifying any other random fields would have no effect

File: db/updateItem.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		Item:   *newName,
+	}
+	itemToInsert, err := dynamodbattribute.MarshalMap(item)
+	if err != nil {
+		return nil, err
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:                itemToInsert,
+		TableName:           aws.String(d.conf.TableNames.Items),
```
[the remark this answers is not in the document]
Comment (samstradling) [reply db/updateItem.go:24]: `Id` is the key name for the table:
https://github.com/mount-joy/thelist-lambda/blob/4c84eb04a9290e547aef9f57e5cafadee6f3a680/cf/1.tables.yml#L20-L24

File: db/updateItem.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		ConditionExpression: aws.String("attribute_exists(Id) AND attribute_exists(ListId)"),
```
[the remark this answers is not in the document]
Comment (samstradling) [reply db/updateItem.go:25]: `Id` is the key name for the table:
https://github.com/mount-joy/thelist-lambda/blob/4c84eb04a9290e547aef9f57e5cafadee6f3a680/cf/1.tables.yml#L20-L24

File: db/updateItem.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (samstradling) [reply db/updateItem.go:35]: According to this the only thing in the error interface is `err.Error()`:
https://blog.golang.org/error-handling-and-go

[the remark this answers is not in the document]
Comment (alexNaylorTeece) [reply db/updateItem.go:35]: Sorry, I missed the brackets - `err.(type)`
I've built an example of what I'm trying to say here: https://play.golang.org/p/zqM53h68Cdz

File: handlers/patchItem.go [hunk 6]
```diff
@@ -41,12 +43,10 @@ func (g *patchItem) handle(request events.APIGatewayV2HTTPRequest) (interface{},
...
 	}
 
-	item, err := g.db.UpdateItem(listID, itemID, newName)
+	item, err := p.db.UpdateItem(listID, itemID, newName)
 	if err != nil {
-		if aerr, ok := err.(db.Error); ok {
-			if aerr.ErrorType == db.ErrorNotFound {
-				return nil, http.StatusNotFound
-			}
+		if errors.Is(err, db.ErrorNotFound) {
```
Comment (alexNaylorTeece) [inline handlers/patchItem.go:48]: Could do a switch here to avoid if in an if, but not crucial

**Pull request merged**